### PR TITLE
Improve performance of Boolean extractors such as `NonFatal`

### PR DESCRIPTION
**Pull request by som-snytt**
As suggested by https://github.com/scala/scala/pull/10600

### CODE REVIEW

**Comment by som-snytt** [comment]
It's a naive rewrite with no tree copying. Curious to hear from lrytz if it's OK. I will give it fresh eyes next sleep cycle. I used this occasion to remember how to write a test.

File: src/reflect/scala/reflect/internal/Definitions.scala [hunk 1]
```diff
@@ -711,7 +711,7 @@ trait Definitions extends api.StandardDefinitions {
     lazy val SuccessClass = requiredClass[scala.util.Success[_]]
     lazy val FutureClass = requiredClass[scala.concurrent.Future[_]]
     lazy val PromiseClass = requiredClass[scala.concurrent.Promise[_]]
-    lazy val NonFatalClass = requiredClass[scala.util.control.NonFatal.type]
+    lazy val NonFatalClass = requiredModule[scala.util.control.NonFatal.type]
```
Comment (lrytz) [inline src/reflect/scala/reflect/internal/Definitions.scala]:
```suggestion
    lazy val NonFatalModule = requiredModule[scala.util.control.NonFatal.type]
```

**Review by lrytz — COMMENTED** [review]
Repurposing the tree seems fine here.

I don't feel strongly about special-casing `NonFatal` to save one `Some` allocation, maybe @SethTisue or @sjrd have an opinion.

**Comment by sjrd** [comment]
No strong feeling either way for me either.

However, in terms of implementation, it seems to me that this should not be done during typer, but rather during a later phase. In dotc it would logically go into a mini-phase. For nsc it could possibly go directly in the pattern matcher.

**Comment by Ichoran** [comment]
When I was benchmarking various error-handling scenarios about a decade ago, using stackless exceptions was noticeably faster when avoiding the allocation.  I no longer remember the details, but it was enough (more than a few percent) so that I have since then never used the extractor.  It's easier to type a few more letters than wonder whether this code might be part of a bottleneck.

Saving people in the future from having to worry seems like a valuable goal, if it's actually this easy.

File: src/reflect/scala/reflect/internal/Definitions.scala (hunk 1 above)
Comment (som-snytt) [reply src/reflect/scala/reflect/internal/Definitions.scala]: that thought crossed my mind. Also I don't know how these members are/were used, and updating the forcing is extra work.

**Comment by som-snytt** [comment]
For the record, I did it at typer for ease of implementation. I had just looked at that code for some reason.

I like Lukas's idea to spec boolean extractor with optional "identity extraction"; possibly that requires early intervention anyway?

**Comment by som-snytt** [comment]
Pushed the naming convention. Those defs are related to async? I'll try to justify the change, in the absence of which the change is "uninteresting".

**Comment by sjrd** [comment]
> I like Lukas's idea to spec boolean extractor with optional "identity extraction"; possibly that requires early intervention anyway?

Yes, that would require typer work. It would also require a SIP, as it's a language change, so it won't happen tomorrow.

**Comment by som-snytt** [comment]
Of course, tomorrow is Saturday.

**Comment by lrytz** [comment]
The `List.apply` rewrite is in CleanUp, it could probably go there.

For the SIP, it seems useful in particular for exception handler extractors. But we couldn't change `NonFatal` anyway due to binary compatibility.

**Comment by som-snytt** [comment]
It's not impossible but it's not pretty. The extra null check is for extraction. The if conditional becomes the `NonFatal.apply`. The body would use x4 directly. It is much brittle work in comparison for a small gain in compartmentalization.
```
  <synthetic> val x4: Throwable = ex6;
  case9(){
    if (x4.ne(null))
      {
        <synthetic> val o11: Option = scala.util.control.NonFatal.unapply(x4);
        if (o11.isEmpty().unary_!())
          {
            val e: Throwable = o11.get().$asInstanceOf[Throwable]();
            matchEnd8({
              e.printStackTrace();
              -1
            })
          }
        else
          case10()
      }
    else
      case10()
  };
```
I will try `Refchecks` first. Really it should go between parser and typer, in a phase that does all sorts of rewrites where we say, parser is doing too much work. In this case, I need to resolve the `NonFatal`, so I want to nudge `Namer`, but no more.

Why is it called `refchecks` instead of `pre-patmat`.

**Comment by som-snytt** [comment]
I got close in refchecks with `transformCaseDefs` but typechecking or manually fixing types is just out of reach.

Why am I spending another minute to move a few lines of code to a later phase that doesn't offer easy ergonomics?

If it's worth doing, it's worth doing as a quasi-syntactic rewrite.

**Comment by lrytz** [comment]
I gave it a try, please take a look. I'm not opposded to doing it in typer out of principle, but I think if we can preserve the original AST for longer it's potentially less breaking / surprising for compiler plugins or other tools that work with typed trees.

**Comment by som-snytt** [comment]
You found it a nice home. It is a patmat concern.

**Comment by lrytz** [comment]
> It is a patmat concern.

Maybe not necessarily, but after patmat it gets a lot more complex.

**Comment by som-snytt** [comment]
... by which I may expand, patmat is supposed to find an efficient representation of whatever patterns and guards we write. So its mission creeps only a bit.

**Comment by som-snytt** [comment]
Thanks @lrytz this is the one where you moved it to patmat.

**Comment by som-snytt** [comment]
> It's a naive rewrite

A bit too naive.

https://github.com/scala/bug/issues/13018

**Pull request merged**